File: Java/ListaExerciciosJavaPOO/Ex002/src/main/java/com/mycompany/ex002/ContaCorrente.java
package com.mycompany.ex002;

public class ContaCorrente {

    private int numeroConta;
    private String nomeCorrentista;
    private float saldo = 0.0f;

    public ContaCorrente(int numeroConta, String nomeCorrentista) {
        this.numeroConta = numeroConta;
        this.nomeCorrentista = nomeCorrentista;
    }

    public ContaCorrente(float saldo, int numeroConta, String nomeCorrentista) {
        this.saldo = saldo;
        this.numeroConta = numeroConta;
        this.nomeCorrentista = nomeCorrentista;
    }

    public void alterarNome(String novoNome) {
        this.nomeCorrentista = novoNome;
    }

    public void deposito(float valorDeposito) {
        try {
            this.saldo += valorDeposito;
            System.out.println("Depósito de R$" + valorDeposito + " efetuado com sucesso.");
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    public void saque(float valorSaque) {
        try {
            if (valorSaque > saldo) {
                System.out.println("Saldo insuficiente.");
            } else {
                this.saldo -= valorSaque;
                System.out.println("Saque de R$" + valorSaque + " efetuado com sucesso.");
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    @Override
    public String toString() {
        return "ContaCorrente{" + "numeroConta=" + numeroConta + ", nomeCorrentista=" + nomeCorrentista + ", saldo=" + saldo + '}';
    }

}

File: Golang/src/banco/contas/contaCorrente.go
package contas

/* Struct é como uma classe
A váriavel com letra minuscula age como private, enquanto com maiuscula como public
*/
import "banco/clientes"

type ContaCorrente struct {
	Titular                    clientes.Titular
	NumeroAgencia, NumeroConta int
	saldo                      float64
}

func (c *ContaCorrente) Sacar(valorSaque float64) string {
	podeSacar := valorSaque <= c.saldo && valorSaque > 0
	if podeSacar {
		c.saldo -= valorSaque
		return "Saque realizado com sucesso"
	} else {
		return "saldo insuficiente"
	}
}

func (c *ContaCorrente) Depositar(valorDeposito float64) (string, float64) {
	if valorDeposito < 0 {
		return "Valor inválido", c.saldo
	}
	c.saldo += valorDeposito
	return "Deposito realizado com sucesso", c.saldo
}

func (c *ContaCorrente) Transferir(valorTransferencia float64, contaDestino *ContaCorrente) string {
	if valorTransferencia < c.saldo && valorTransferencia > 0 {
		c.Sacar(valorTransferencia)
		contaDestino.Depositar(valorTransferencia)
		return "Transferência realizada com sucesso"
	}
	return "Saldo insuficiente"
}

func (c *ContaCorrente) Extrato() float64 {
	return c.saldo
}
